Add Plugin: Prayer Stats / Dose Indicator

shmeeps opened the pull request:
Adds a prayer stats plugin. 

Default functionality displays a glowing border around the quick prayer orb when a player has a prayer potion or super restore potion in their inventory, and are missing enough prayer points to receive the full prayer restore value from a dose of that potion.

![ezgif-1-43c09ad814](https://user-images.githubusercontent.com/1050677/40398156-22f85136-5dfc-11e8-8042-f87691037568.gif)

In addition, a tool-tip is enabled when hovering over the quick-prayer orb that can show the estimated time remaining until your prayer points run out, as well as your current prayer bonus. Color indicators are configurable.

Addresses #598. Provides some feature parity with dosing indicators from OSBuddy.

First attempt at a plugin, so appreciate any feedback.

Reasel commented:
This might look a lot better as an infobox that was displayed only when you can drink either a restore or a pray pot.

shmeeps commented:
@Reasel All the lines are toggle-able, and it can be configured to only display the notifications when a dose is required. It will also pick up and display only when you have a prayer potion and/or super restore in your inventory, by default.

Infinitay commented:
I would suggest changing the infobox's background color to red when your prayer is low. Heck, yellow when 25-50 % and red <25% would be nice

shmeeps commented:
@Infinitay Great idea, I'll look into it.

Adam- commented:
Also we get a lot of requests for infoboxes showing what prayers you have on (mostly non-overheads like piety)

Adam- commented:
Actually this is related to #2840

deathbeam commented:
This should be probably just tooltip on the prayer icon... . Also, I really liked the idea of glowing prayer icon when there is need to drink.

shmeeps commented:
Changed up the functionality to display as infobox icons by default, which include a number to indicate how many doses of each type of potion that are available can be sipped. 

The stats panel can be enabled to display the prayer points remaining, prayer bonus, and remaining time details, as well as the dose indicators and the holy wrench status which are disabled by default.

deathbeam commented:
I feel like you are trying to do too much at once. there is no need to display the dose indicators, for feature parity with OSB all that is needed is the glow around prayer orb. Info for everything else you have in your inventory. Also, if you really want to do the informative tooltip, as i said earlier, you can just do it as tooltip on the in-game prayer orb.

shmeeps commented:
Updated to convert the indicator to a strobing border around the quick-prayer orb. Colors are configurable. Demo in the OP.

The time remaining and prayer bonus now show in a tool-tip when hovering over the quick-prayer orb. The rest of the tool-tip data was removed, as it is superfluous when looking at the quick-prayer orb.

deathbeam commented:
Looks great, ty for revisiting it. Umm, not sure how the color chooser is really useful (for the first to second color changes for warning), it seems like reasonable default, i dont think anyone would need/want to change it anyway.

Kamielvf commented:
I would just copy what was done over at #1578 regarding the colours + the same flashing speed (make it feel more like a glow) as the one over at #1578 is nicer to look at and much less intrusive.

shmeeps commented:
Updated to blue colors by default, and changed the pulse to complete a cycle every 2 ticks instead of everyone 1 tick. Demo above.

Infinitay commented:
You should not be using wildcard imports (https://github.com/runelite/runelite/wiki/Code-Conventions)

Few suggestions:
 - Maybe move your `PrayerData` changes into the already existing `Prayer` api (in runelite-api)
 - Make use of Lombok's `@Getter` and `@Setter` instead of creating getters/setters for multiple variables (`PrayerDosePlugin`)

shmeeps commented:
Fixed the auto imports, moved the prayer drain rates into the prayer API, converted getters/setters to lombock, and refactored the helper class into the plugin class to reduce the redundancy of code from an earlier implementation.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/praydose/PrayerDoseConfig.java:
```diff
@@ -0,0 +1,83 @@
+/*
```
Add correct license header

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/praydose/PrayerDoseOverlay.java:
```diff
@@ -0,0 +1,188 @@
+/*
```
Add correct license header (your name)

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/praydose/PrayerDosePlugin.java:
```diff
@@ -0,0 +1,283 @@
+/*
```
Add correct license header

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/praydose/PrayerItems.java:
```diff
@@ -0,0 +1,443 @@
+package net.runelite.client.plugins.praydose;
```
Add license header, fix imports (https://github.com/runelite/runelite/wiki/Code-Conventions)

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/praydose/PrayerItems.java:
```diff
@@ -0,0 +1,443 @@
+		}
+
+		// Blessings
+		PrayerItems.itemToPrayerBonus.put(ItemID.PEACEFUL_BLESSING, 1);
```
Why are you referencing class name here, not needed

deathbeam replied:
Also, to follow existing conventions, I would make this entire thing Enum

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/praydose/PrayerDosePlugin.java:
```diff
@@ -0,0 +1,283 @@
+			return;
+		}
+
+		this.checkInventory();
```
Instead of checkEquipment and checkInventory each game tick, just use ItemContainerChanged event, basically, remove this entire gameTick hook and use proper events

shmeeps commented:
Updated licenses and imports, converted the PrayerItems class into an enum, and converted the tick inventory check to onItemContainerChanged.

shmeeps commented on runelite-client/src/main/java/net/runelite/client/plugins/praydose/PrayerDosePlugin.java:
```diff
@@ -112,16 +115,15 @@ public Overlay getOverlay()
-			return;
-		}
-
 		this.checkInventory();
```
Are the events atomic or are they like grouped by tick and subject to race conditions? Slightly concerned there might be some de-sync if you equipped a ring of the gods, for example, and the equipment container event ran before the inventory container event. The first would pick up the holy wrench in the equipment, but the second would not pick up anything in your inventory, and that would throw off the prayer calculations. I think I can just split the holy wrench flag into two pieces to indicate whether it's present in your inventory or equipment, but I'm not sure if that's necessary.

deathbeam replied:
Grouped by tick

deathbeam replied:
You should at least check if this is either inventory or equipment and then continue, because otherwise you will get million of events when opening bank

shmeeps replied:
Got it. Changed it to only process equipment and inventory events. Also went ahead and split up the two holy wrench statuses / container checks, so it's a bit less processing on each event. Unfortunately don't have an equipable holy wrench item to test with, but I can't forsee any issues with it.

shmeeps commented:
Conflicts resolved

deathbeam commented on runelite-api/src/main/java/net/runelite/api/Prayer.java:
```diff
@@ -32,125 +32,128 @@
-	Prayer(Varbits varbit)
+	private final double drainRate;
+
+	private Prayer(Varbits varbit, double drainRate)
```
access modifier for enum constructor is unnecessary

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/praydose/PrayerDoseConfig.java:
```diff
@@ -0,0 +1,82 @@
+		name = "Dose Color 1",
+		description = "First color displayed when a dose is available and needed."
+	)
+	default Color doseColor1()
```
These 2 color options are unnecessary

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/praydose/PrayerDoseOverlay.java:
```diff
@@ -0,0 +1,187 @@
+	private boolean trackTick = true;
+
+	@Inject
+	public PrayerDoseOverlay(Client client, PrayerDosePlugin plugin, PrayerDoseConfig config)
```
Make this constructor private

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/praydose/PrayerDoseOverlay.java:
```diff
@@ -0,0 +1,187 @@
+	private final PrayerDoseConfig config;
+	private final int tooltipOffset = 45;
+	private final float pulseTime = 1200f;
+	private Instant startOfLastTick = Instant.now();
```
Move this 1 line below, group final values together

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/praydose/PrayerDoseOverlay.java:
```diff
@@ -0,0 +1,187 @@
+	private final PrayerDosePlugin plugin;
+	private final PrayerDoseConfig config;
+	private final int tooltipOffset = 45;
+	private final float pulseTime = 1200f;
```
tooltipOffset and pulseTime looks like constants? make them private static final and UPPER_CASE

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/praydose/PrayerDoseOverlay.java:
```diff
@@ -0,0 +1,187 @@
+		return new Dimension((int) bounds.getWidth(), (int) bounds.getHeight());
+	}
+
+	private void renderTooltip(Graphics2D graphics, Rectangle2D bounds)
```
You can probably just use TooltipManager here

shmeeps commented:
Update code issues, converted to TooltipManager

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/praydose/PrayerDoseConfig.java:
```diff
@@ -0,0 +1,59 @@
+)
+public interface PrayerDoseConfig extends Config
+{
+	@ConfigItem(
```
This configuration is probably not necessary, its tooltip after all. I would just add general toggle for disabling the tooltip

shmeeps commented:
Conflicts resolved and rebased.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/praydose/PrayerDoseOverlay.java:
```diff
@@ -0,0 +1,173 @@
+	private final Client client;
+	private final PrayerDosePlugin plugin;
+	private final PrayerDoseConfig config;
+	private final PanelComponent panelComponent = new PanelComponent();
```
this looks unused now

shmeeps replied:
Removed.

shmeeps commented:
@deathbeam I'm considering merging this plugin with the existing prayer plugin, as that one now contains both the flick timer and the prayer info boxes, and it would help reduce plugin clutter. Thoughts?

deathbeam commented:
Yeah probably good idea.

Adam- commented:
I agree this is a good idea.

shmeeps commented:
Done, now a part of the prayer plugin.
